pkg/agent/multicluster/regular_node_route_controller.go

```diff
@@ -0,0 +1,412 @@
+// Copyright 2022 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package noderoute
+
+import (
+	"fmt"
+	"net"
+	"strings"
+	"time"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+
+	mcv1alpha1 "antrea.io/antrea/multicluster/apis/multicluster/v1alpha1"
+	mcclientset "antrea.io/antrea/multicluster/pkg/client/clientset/versioned"
+	mcinformers "antrea.io/antrea/multicluster/pkg/client/informers/externalversions/multicluster/v1alpha1"
+	mclisters "antrea.io/antrea/multicluster/pkg/client/listers/multicluster/v1alpha1"
+	"antrea.io/antrea/pkg/agent/config"
+	"antrea.io/antrea/pkg/agent/interfacestore"
+	"antrea.io/antrea/pkg/agent/openflow"
+	"antrea.io/antrea/pkg/ovs/ovsconfig"
+)
+
+const (
+	controllerName = "AntreaAgentMCNodeRouteController"
```

Comment:
Do we really need two controllers one for GW one for other Nodes, which watch / handle the same events? Could you explain why that is better?

There can be no conflicts of the flows added by the two controllers (note they are totally async and add flows without atomicity), that can lead to unexpected/non-deterministic forwarding behaviors?

Reply:
Yeah, I thought that a little bit before, personally I feel there will be lots of if/else to check it's Gateway or regular Node to run corresponding codes which is not so readable. Considering there might be more different behaviors for Gateway and Node in the future, so I choose to use two controllers.

Reply:
I will check and a few more share functions to reduce duplicate codes between two controllers.

Reply:
I doubt that really helps. In the worst case, you can have one controller, but for every event call two functions, one for GW and one for regular Node (not saying we should do that).

Reply:
Refactored the codes and now is only one controller for both regular and Gateway Nodes.